packages/workspace/src/browser/workspace-service.ts

```diff
@@ -57,18 +66,128 @@ export class WorkspaceService implements FrontendApplicationContribution {
 
     @postConstruct()
     protected async init(): Promise<void> {
-        const rootUri = await this.server.getRoot();
-        this._root = await this.toValidRoot(rootUri);
-        if (this._root) {
-            const uri = new URI(this._root.uri);
-            this.updateTitle(uri);
-            this.watcher.watchFileChanges(uri);
+        this._workspaceDescription = {
+            root: undefined,
+            roots: [],
+            workspaceConfigFile: undefined
+        };
+        await this.updateWorkspace();
+        this.updateTitle();
+        this.deferredWorkspaceDescription.resolve(this._workspaceDescription);
+        const configUri = this.getWorkspaceConfigFileUri();
+        if (configUri) {
+            const configUriString = configUri.toString();
+            this.watcher.onFilesChanged(changes => {
+                if (changes.some(change => change.uri.toString() === configUriString)) {
+                    this.updateWorkspace();
+                }
+            });
+        }
+    }
+
+    get root(): Promise<FileStat | undefined> {
+        return this.deferredRoot.promise;
+    }
+    get workspace(): Promise<WorkspaceDescription> {
+        return this.deferredWorkspaceDescription.promise;
+    }
+    get numberOfRoots(): number {
+        return this._workspaceDescription.roots.length;
+    }
+
+    protected readonly onWorkspaceChangeEmitter = new Emitter<WorkspaceDescription>();
+    get onWorkspaceChanged(): Event<WorkspaceDescription> {
+        return this.onWorkspaceChangeEmitter.event;
+    }
+
+    protected async updateWorkspace(): Promise<void> {
+        if (!this._root) {
+            const rootUri = await this.server.getWorkspace();
+            this._root = await this.toValidRoot(rootUri);
         }
+
         this.deferredRoot.resolve(this._root);
```

Comment:
It should not be resolved here and not with the workspace description, but later with the real workspace roots.